### PR TITLE
Do not overwrite existed image layer

### PR DESCRIPTION
**Pull request by knizhnik**
refer #1594

refer #1690

### CODE REVIEW

**Review by dhammika — APPROVED** [review]
Nice!! guess the root cause was compaction get(key) triggering get_reconstruct_data(key) from the same layer image. https://github.com/neondatabase/neon/blob/b9f84f4a83ed916919884b4f9f038356e76f113f/pageserver/src/layered_repository.rs#L1913-L1920
I'd suggest we check the layer map instead of IO error. In case we have some stale layer image that's not in the layer map.

**Comment by hlinnaka** [comment]
I don't understand. When would we try to create an image file that already exists? After a crash and restart?

**Comment by hlinnaka** [comment]
>  I'd suggest we check the layer map instead of IO error. In case we have some stale layer image that's not in the layer map.

+1

**Comment by knizhnik** [comment]
> I don't understand. When would we try to create an image file that already exists? After a crash and restart?

No, just two subsequent invocation of GC when no changes made in the database (LSN is not advanced).
In this case image layer creation (which s part of compaction may try to rewrite existed image).

**Comment by knizhnik** [comment]
> > I'd suggest we check the layer map instead of IO error. In case we have some stale layer image that's not in the layer map.
> 
> +1

May be checking layer map is better than handling AlreadyExists error.
Just want to notice:
1. There is currently no method in layer map which can be used to check presence of layer with exact key/lsn. `search` and `find_latest_image` do more complicated task. Certainly it is possible to use them, but I think it will be better to add special method for exact match.
2. Image layer file should be opened with create_new(true) to throw error if file already exists.

**Comment by hlinnaka** [comment]
>     1. There is currently no method in layer map which can be used to check presence of layer with exact key/lsn. `search` and `find_latest_image` do more complicated task. Certainly it is possible to use them, but I think it will be better to add special method for exact match.

+1 for adding a new method for exact match.

>     2. Image layer file should be opened with create_new(true) to throw error if file already exists.

+1

**Comment by hlinnaka** [comment]
> No, just two subsequent invocation of GC when no changes were made in the database (LSN is not advanced).
> In this case image layer creation (which is part of compaction) may try to rewrite existed image layer.

Hmm, I still don't understand how that would happen. If the first GC iteration creates a set of image files, the `time_for_new_image_layer` call on the next iteration would return false. What am I missing?

To be clear, I agree we should make this more robust if it does happen. And I believe it is happening for some reason. But I'd still like to understand the exact steps how it happens.

**Comment by knizhnik** [comment]
> > No, just two subsequent invocation of GC when no changes were made in the database (LSN is not advanced).
> > In this case image layer creation (which is part of compaction) may try to rewrite existed image layer.
> 
> Hmm, I still don't understand how that would happen. If the first GC iteration creates a set of image files, the `time_for_new_image_layer` call on the next iteration would return false. What am I missing?
> 
> To be clear, I agree we should make this more robust if it does happen. And I believe it is happening for some reason. But I'd still like to understand the exact steps how it happens.

Yes, I agree with you. My patch solves the consequence of the problem, but not the source of the problem. And the source of the problem is why we try to create new image layer if previous one exists and there are no new changes.
Looks like something in image_coverage/time_for_new_image_layer is not working correctly.

**Comment by knizhnik** [comment]
The problem was in incorrect LSN range in time_for_new_image_layer (end > start).

**Comment by knizhnik** [comment]
@hlinnaka I have revisited this patch: remove check for already exists error and just add check for valid LSN range.
May be instead we should change `range_overlaps`. Right now `range_overlaps(&(0..3), &(2..1)) == true`.

File: pageserver/src/layered_repository.rs [hunk 1]
```diff
@@ -2133,15 +2134,18 @@ impl LayeredTimeline {
...
                         debug!("future({})", lsn);
                         pitr_cutoff_lsn = cutoff;
                     }
-                    LsnForTimestamp::Past(lsn) => {
+                    Ok(LsnForTimestamp::Past(lsn)) => {
                         debug!("past({})", lsn);
                     }
+                    Err(_) => {
+                        debug!("no commits found");
+                    }
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:2148]: Why this change?

File: pageserver/src/layered_repository/image_layer.rs [hunk 2]
```diff
@@ -430,7 +430,10 @@ impl ImageLayerWriter {
             },
         );
         info!("new image layer {}", path.display());
-        let mut file = VirtualFile::create(&path)?;
+        let mut file = VirtualFile::open_with_options(
```
Comment (hlinnaka) [inline pageserver/src/layered_repository/image_layer.rs:433]: Update the comment above

File: pageserver/src/layered_repository.rs [hunk 3]
```diff
@@ -1888,15 +1888,16 @@ impl LayeredTimeline {
                 } else {
                     Lsn(0)
                 };
+                if img_lsn < lsn {
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:1891]: I don't understand this change. `image_coverage` above shouldn't return images newer than 'lsn'. Is the point of this check to ignore images with `img_lsn == lsn`? For such a case, `count_deltas` will surely return 0.

I'm sure this logic could use some refactoring and extra comments, but I don't think this is an improvement.

File: pageserver/src/layered_repository.rs (hunk 1 above)
Comment (knizhnik) [reply pageserver/src/layered_repository.rs:2148]: It cause failures in test_branch_and_async_gc at the beginning: find_lsn_for_timestamp returns error if there are no transactions and it can not locate timestamp.

File: pageserver/src/layered_repository.rs (hunk 3 above)
Comment (knizhnik) [reply pageserver/src/layered_repository.rs:1891]: > I don't understand this change. `image_coverage` above shouldn't return images newer than 'lsn'. Is the point of this check to ignore images with `img_lsn == lsn`? For such a case, `count_deltas` will surely return 0.
> 
> I'm sure this logic could use some refactoring and extra comments, but I don't think this is an improvement.

This is not an improvement: this is actually bug fix. I just investigated why new image layer is created if previous one exists and find out that `time_for_new_image_layer`  returns true despite to the fact that there were no changes  since last page image.  And intervals check for overlap in `count_deltas`:
```
            if !range_overlaps(&l.get_lsn_range(), lsn_range) {
                continue;
            }
```
really looks like range_overlaps(&(0..3), &(2..1)) 
LSN of image is 1.
So  overlaps returns `true`, `count_deltas` - 1 and as far as 1 is compaction threashold in this test we try to create new image layer with the same LSN and so overwrite existed layer.

Comment (hlinnaka) [reply pageserver/src/layered_repository.rs:1891]: Hmm, I see. Let me walk through an example to get this written down:

In the beginning, there are three delta layers (as an example):

delta layer with LSN range 71-81
delta layer with LSN range 81-91
delta layer with LSN range 91-101

`time_for_new_image_layer` is called with LSN 100, which is the last_record_lsn. It calls count_deltas(0..100), which returns 3, because there is no previous image layer. A new image layer is created at LSN 100.

Now we have:
delta layer with LSN range 71-81
delta layer with LSN range 81-91
delta layer with LSN range 91-101
image layer at LSN 100

Some new WAL records are ingested, so that last_record LSN is 105, and `time_for_new_image_layer` is called again. It calls count_deltas(img_lsn..lsn), which means `count_deltas(100..105)`. count_deltas returns 1, because it considers `91..101` to overlap with `100..105`.

I'm not entirely sure what the best fix would be. It feels wrong that count_deltas() returns 1 in this case, there aren't really any new delta layers on top of the image layer. So perhaps we should call count_deltas as `count_deltas(img_lsn + 1, lsn)`.

Note: There's another twist here. In the above example img_lsn is *not* equal to lsn, because some new WAL was ingested. So adding the check for `img_lsn > lsn` would not prevent a new image layer from being created. I'm not sure if you did that intentionally. It seems like a waste of time to create a new image layer, if there are no deltas after the previous image layer. The new image layer would have the same contents as the old one. Then again, if we define the threshold of 1 to mean that you create a new image even if there's been no changes since last one, as long as the LSN would be different, then it's correct. It doesn't sound right to me though.

If no new WAL was ingested, then count_deltas would be called with (100..100). It would still return 1, because it considers `100..100` to overlap with `91..101`. I'm not sure if that's sensible; can an empty range overlap with anything? Your explicit check for `img_lsn > lsn` would prevent the new image layer from being created in this case, however.

So what's the best fix? I think we should do two things:

1. Call count_deltas with range `img_lsn+1..lsn`. Because if there are no WAL records after the last image, we want num_deltas to be 0. 
2. Add the explicit check for `img_lsn > lsn`. It's not strictly necessary if we change the `count_deltas` call, but in general, it never seems like a good idea to create a new image layer with exactly the same LSN that the last image layer had.


Off-by-one bugs are tricky..

Comment (knizhnik) [reply pageserver/src/layered_repository.rs:1891]: > Count_deltas returns 1, because it considers 91..101 to overlap with 100..105.

Not exactly. Please notice that in my case img_lsn>lsn.
So actually 'count deltas` is called for LSN range img_lsn..lsn=101..100 which is assumed to overlap with 100..105.
How it is possible? `lsn` is taken for the particular partition. If this partition is not changed then it is equal to image lsn=100.

I do not understand how replacing `img_lsn..lsn` with  `img_lsn+1..lsn`. We  will get 102..100 instead of 101..100, but it still assumed to overlap with  100..105.

Comment (hlinnaka) [reply pageserver/src/layered_repository.rs:1891]: > Please notice that in my case img_lsn>lsn

How does that happen? AFAICS, image_coverage() cannot return images layers with img_lsn > lsn.

(one lesson here is that we should write unit tests for the functions in layer_map.rs)

Comment (knizhnik) [reply pageserver/src/layered_repository.rs:1891]: > How does that happen? AFAICS, image_coverage() cannot return images layers with img_lsn > lsn.

image_coverage returns last_img with LSN=100.
Then 
```
   let img_lsn = if let Some(last_img) = last_img {
                    last_img.get_lsn_range().end
```
and  `last_img.get_lsn_range().end` is 101

Comment (hlinnaka) [reply pageserver/src/layered_repository.rs:1891]: Oh, I see. But hang on, now I'm confused again. If you call count_deltas(101..100), it returns 0. Right?

It would be nice to refactor this logic to a separate struct or module that we could write unit tests for. That would allow us to write test functions to demonstrate all these corner cases. Maybe not in this PR, but then again, maybe it would be easier to just do it now. It's hard to write tests for all these corner cases otherwise.

Comment (knizhnik) [reply pageserver/src/layered_repository.rs:1891]: > If you call count_deltas(101..100), it returns 0. Right?

No. As I wrote above `range_overlaps(&(100..105), &(101..100))` returns `true` and `count_deltas(&(101..100))` returns 1.
This is why new image layers is created with the same coordinates as old one.
The problem is easily reproduce with test from #1690. To analyze  the situation when it happens I recommend to replace `VirtualFile::create(&path)?` in  image_layer.rs with 
```
let mut file = VirtualFile::open_with_options(
            &path,
            std::fs::OpenOptions::new().write(true).create_new(true),
        ).unwrap();
```

Comment (hlinnaka) [reply pageserver/src/layered_repository.rs:1891]: > > If you call count_deltas(101..100), it returns 0. Right?
> 
> No. As I wrote above `range_overlaps(&(100..105), &(101..100))` returns `true` and `count_deltas(&(101..100))` returns 1.

In this example, there is no layer 100..105. So the call would be to `rang_overlaps(&(91..100), &(101..100)))`, which returns false.

Hmm, there is a race condition here though, so that it *is* possible to have a layer 100..105, even though lsn=100. That's because we read last_record_lsn first, and call count_deltas() later. last_record_lsn can advance, and new layers can be created between the calls to get_last_record_lsn() and time_for_new_image_layer(). Is that what happened here?

Comment (knizhnik) [reply pageserver/src/layered_repository.rs:1891]: > Is that what happened here?

Looks like. At least I do not see any other explanation why we may want to create new image layer with the same LSN as old image layer, while delta layer inspected by count_deltas has larger LSN.

Comment (knizhnik) [reply pageserver/src/layered_repository.rs:1891]: So what do you think about just adding this check (`img_lsn < lsn`).
Race condition in count_deltas seems to be not critical if we  avoid creation of image layer with the same LSN.
And this check prevents it.

**Review by hlinnaka — APPROVED** [review]

**Comment by hlinnaka** [comment]
Rebased and squashed

File: pageserver/src/layered_repository.rs (hunk 1 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository.rs:2208]: After this discussion here, commit f09c09438a307f88eeb6bbaf85069022623e35d6 added a `LsnForTimestamp::Nodata` case for this, so I believe we don't need to do anything about it in this PR anymore.

**Comment by hlinnaka** [comment]
Rebased again. If the CI is happy with this, I'll push it later today.
